k8s/sonarqube/README.md

```diff
@@ -0,0 +1,462 @@
+  - A web server allowing to configure SonarQube instance
+  - A search server which is based on Elasticsearch
+  - A compute engine server for analysis and processing
+- SonarQube Database
+- A set of plugins
+- SonarScanners which are responsible for project analysis - those usually run on build servers
+For our application database the most important place, `plugins` folder stored on PVC (Persistent Volume Claim).
+
+## Backing up plugin and data
+
+For backup pluging you have to copy files from folder `$SONARQUBE_HOME/extension`.
+This shell script will create copy of plugins folder in current folder
+
+```shell
+kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
+                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/extensions/ ./
+```
+
+This shell script will create copy of content from folder `data` in current folder:
+
+```shell
+kubectl --namespace $NAMESPACE cp $(kubectl -n$NAMESPACE get pod -oname | \
+                sed -n /\\/$APP_INSTANCE_NAME-sonarqube/s.pods\\?/..p):/opt/sonarqube/data ./
+```
+
+it is not necessary to backup data folder it will be downlnoaded by application
+from database.
+
+## Backing up PostgreSQL
+
+All configuration and data about projects stored in database.
+This shell script will create `postgresql-backup.sql` dump of all DB in PostgreSQL.
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+	$(kubectl -n$NAMESPACE get pod -oname | \
+		sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+	-c postgresql-server \
+	-- pg_dumpall -c -U postgres > backup.sql
+```
+
+## Restoring
+
+Entering to maintenance mode:
+In order to restore PostgresSQL database there is a need to perform some preliminary steps:
+The command below will “lock” the database by blocking new incoming connections:
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+     sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres -c "REVOKE CONNECT ON DATABASE sonar FROM PUBLIC;"
+```
+
+Next in order to ensure data consistency all active connections will be dropped:
+
+```shell
+kubectl --namespace $NAMESPACE exec -t \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+     sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
+```
+
+Now you are able to restore data to the database.
+Below shell script will restore data from `backup.sql` to PostgreSQL
+
+```shell
+cat backup.sql | kubectl --namespace $NAMESPACE exec -i \
+  $(kubectl -n$NAMESPACE get pod -oname | \
+    sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
+  -c postgresql-server \
+  -- psql -U postgres
+```
+
+Next copy files from current folder to folder `data` in SonarQube application pod:
```

Comment:
Instead of writing next, afterwards prefixes. Could you use a list?

For example:


1.  Run the command below to **lock** the database by blocking new incoming connections:

    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
       -c postgresql-server \
          -- psql -U postgres -c "REVOKE CONNECT ON DATABASE sonar FROM PUBLIC;"
    ```

1.  Ensure data consistency, all active connections will be dropped:


    ```shell
    kubectl --namespace $NAMESPACE exec -t \
      $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
       -c postgresql-server \
          -- psql -U postgres -c "select pg_terminate_backend(pid) from pg_stat_activity where datname='sonar';"
    ```

1.  Restore data to the database. Below command restores data from backup.sql to PostgreSQL:

    ```shell
    cat backup.sql | kubectl --namespace $NAMESPACE exec -i \
      $(kubectl -n$NAMESPACE get pod -oname | \
      sed -n /\\/$APP_INSTANCE_NAME-postgresql-deployment/s.pods\\?/..p) \
       -c postgresql-server \
          -- psql -U postgres
    ```

Reply:
fixed